crates/cli/src/native_query/type_solver/simplify.rs

```diff
@@ -574,4 +603,137 @@ mod tests {
+            [TypeConstraint::Union(
+                [
+                    TypeConstraint::Scalar(BsonScalarType::Int),
+                    TypeConstraint::Scalar(BsonScalarType::Null),
+                ]
+                .into(),
+            )],
+        );
+        expect_that!(
+            result,
+            ok(eq(&BTreeSet::from([TypeConstraint::Union(
+                [
+                    TypeConstraint::Scalar(BsonScalarType::Int),
+                    TypeConstraint::Scalar(BsonScalarType::Null),
+                ]
+                .into(),
+            )])))
+        );
+        Ok(())
+    }
+
+    #[googletest::test]
+    fn simplifies_from_nullable_to_non_nullable_in_contravariant_context() -> Result<()> {
+        let configuration = mflix_config();
+        let result = super::simplify_constraints(
+            &configuration,
+            &Default::default(),
+            &mut Default::default(),
+            Some(TypeVariable::new(1, Variance::Contravariant)),
+            [
+                TypeConstraint::Scalar(BsonScalarType::String),
+                TypeConstraint::Union(
+                    [
+                        TypeConstraint::Scalar(BsonScalarType::String),
+                        TypeConstraint::Scalar(BsonScalarType::Null),
+                    ]
+                    .into(),
+                ),
+            ],
+        );
+        expect_that!(
+            result,
+            ok(eq(&BTreeSet::from([TypeConstraint::Scalar(
+                BsonScalarType::String
+            )])))
+        );
+        Ok(())
+    }
+
+    #[googletest::test]
+    fn emits_error_if_scalar_is_not_compatible_with_any_union_branch() -> Result<()> {
+        let configuration = mflix_config();
+        let result = super::simplify_constraints(
+            &configuration,
+            &Default::default(),
+            &mut Default::default(),
+            Some(TypeVariable::new(1, Variance::Contravariant)),
+            [
+                TypeConstraint::Scalar(BsonScalarType::Decimal),
+                TypeConstraint::Union(
+                    [
+                        TypeConstraint::Scalar(BsonScalarType::String),
+                        TypeConstraint::Scalar(BsonScalarType::Null),
+                    ]
+                    .into(),
+                ),
+            ],
+        );
+        expect_that!(
+            result,
+            err(unordered_elements_are![
+                eq(&Error::TypeMismatch {
+                    context: None,
+                    a: TypeConstraint::Scalar(BsonScalarType::Decimal),
+                    b: TypeConstraint::Scalar(BsonScalarType::String),
+                }),
+                eq(&Error::TypeMismatch {
+                    context: None,
+                    a: TypeConstraint::Scalar(BsonScalarType::Decimal),
+                    b: TypeConstraint::Scalar(BsonScalarType::Null),
+                }),
+            ])
+        );
+        Ok(())
+    }
+
+    // TODO:
```

Comment:
Unsupported right now?

Reply:
Yes - I wrote the test to see if it works, and it doesn't. I want to fix that later so I want to keep the test, but I don't want to commit a failing test.